### ISSUE TITLE
force full re-clippy

### ISSUE THREAD

**Issue by davemilter** [rust-lang/rust-clippy#2604]
I build clippy from git sources (4ef7238b3e6b45596a32eb669a07d85d83ff0f1c).
This is re-open of #2498 as @oli-obk suggested.
And actully nothing changes because of  #2582 , and why it should?

Sample code:

```Rust
use std::io;

pub fn map_err(err: io::Error) -> String {
    format!("{}", err)
}
```

```
$ cargo clippy
   Compiling foo v0.1.0 (file:///tmp/foo)
warning: this argument is passed by value, but not consumed in the function body
 --> src/lib.rs:3:21
  |
3 | pub fn map_err(err: io::Error) -> String {
  |                     ^^^^^^^^^ help: consider taking a reference instead: `&io::Error`
  |
  = note: #[warn(needless_pass_by_value)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.191/index.html#needless_pass_by_value

    Finished dev [unoptimized + debuginfo] target(s) in 0.14 secs
$ cargo clippy
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
```

As you use the second run of `clippy` gives nothing, and `cargo clippy --help` 
have no options to force rerun.

**Comment by matthiaskrgr** [comment]
I usually work around this by runnning something like ````touch src/main.rs ; cargo clippy````

**Comment by oli-obk** [comment]
can you try with clippy 0.0.203? I included a possible fix there

**Comment by CyrilCalmels** [comment]
I still have the problem with clippy 0.0.203 :cry:

**Comment by baumanj** [comment]
Also seeing this with v0.0.204

**Comment by Manishearth** [comment]
cc @alexcrichton is there any way we can get a mode to cargo check that always rebuilds the toplevel crate (i.e. the root, or if running -p, that crate). Preferably not clobbering artefacts unless they need to be.

**Comment by dwijnand** [comment]
@killercup is fixing the same bug that `cargo fix` has over in https://github.com/rust-lang/cargo/pull/5750, which I'm hopeful will make fixing this much easier.

**Comment by killercup** [comment]
Memo to self (and promise to clippy maintainers) to actually expose this
somehow to non-cargo consumers!

**Comment by JoshMcguigan** [comment]
rust-lang/cargo#5944 was merged, which is a rebase of killercup's work. I'd be open to working on a fix for this, if there is anyone who can provide guidance on what is needed from the clippy side?

**Comment by matthiaskrgr** [comment]
If the cargo folks agree, perhaps we could have commandline arg in cargo which forces the package to be rebuilt. (cargo check -f / --force or something like that).
I guess the flag then needs to be hooked up in.
https://github.com/rust-lang-nursery/rust-clippy/blob/master/src/main.rs#L103-L111

We would somehow need to make sure it only forces recheck of the crates in the current workspace and not all dependencies as well, not sure what is the best way to do this though.. :/

**Comment by BatmanAoD** [comment]
Even something as simple as detecting that no work has been done and outputting a warning suggesting the user `touch` the root file would improve the user experience drastically, I think.

I believe the ideal solution is a diagnostic-caching mechanism, but currently users have *no* indication that `cargo build && cargo clippy` is a misuse of Clippy, or that their code may have lint warnings or errors that they're not seeing.

**Comment by JoshMcguigan** [comment]
I just submitted rust-lang/cargo#6664 to cargo to force `cargo check` to rebuild. There may be some additional work needed there, but hopefully that will move this issue forward.

**Comment by JoshMcguigan** [comment]
It seems rust-lang/cargo#6664 is close to getting merged. It would be helpful if someone familiar with how clippy integrates with cargo could check in and confirm this would solve the clippy use case.

**Comment by Manishearth** [comment]
Yes, but we may need to pass a flag down

**Comment by crlf0710** [comment]
Encountered this problem - when i run clippy twice with different parameters, the problem didn't get rebuilt, so i got no output.

**Comment by ehuss** [comment]
This is fixed on latest master of cargo for `clippy-preview` (via rust-lang/cargo#7157).  I'll be updating nightly soon.  Migrating to clippy-preview is tracked in #3837.

**Comment by Livven** [comment]
Still an issue for me.

```
cargo clippy --version
clippy 0.0.212 (4ee12063 2020-02-01)

cargo --version
cargo 1.42.0 (86334295e 2020-01-31)

rustc --version
rustc 1.42.0 (b8cedc004 2020-03-09)
```

**Comment by ehuss** [comment]
@Livven It was fixed as a nightly-only feature, which has since been removed.  We are migrating to a new approach, which should hopefully land soonish via #5363.

**Comment by Livven** [comment]
@ehuss Thanks for the info, looking forward to it. Seems like it should also fix #5199.

**Comment by AndreaOrru** [comment]
As of today, I'm still experiencing this issue.
